### ISSUE TITLE
Business rules for submitting responses.

### ISSUE THREAD

**Issue by SHeffernan** [SHeffernan/gploop#27]
Doctors may complete a survey for a given product multiple times during a year, but only once each day. each response needs to be time stamped in order to differentiate responses from the same dr for the same product.

Submission is permanent.

This allows us to ensure that responses are generated during the working day, or at an evening meeting. This would be a security step.

**Comment by pwhipp** [comment]
- If questionnaire is incomplete, take the user to the question that is not yet completed when visited.

- When starting a new questionnaire, check most recent response to the questionnaire and if it is the same date as the current date, inform the user that they have already done the survey today.

- Add an 'Abandon' button allowing the user to abandon the partially completed questionnaire and start again.

Sound good?

**Comment by SHeffernan** [comment]
Yes.

 

BTW: Point 2 will be that they have already completed their feedback for that product today.

It maybe that they complete the same feedback (survey) questions for a different product – which would be Ok.

 

With point 1, how viable is it to highlight the earliest missing response, so they know exactly where to begin again??

**Comment by pwhipp** [comment]
Note that the percentage complete of a questionnaire can go down when a question is answered because we have questions with prerequisites. For example you might have three questions whose prerequisites become met when you answer some other question in a certain way resulting in three questions effectively being added to the questionnaire for that particular user. I've handled this in the logic but mention it here in case it confuses anyone (else).

**Comment by SHeffernan** [comment]
Thanks Paul. 
That makes sense. 

Sean Heffernan 
0438 289 214

**Comment by pwhipp** [comment]
All done. Much testing needed but I think the functionality is ready to populate and use. I've added issue #31 for general feedback which will take an hour or so.

This is a substantial enhancement to the model. For testing I've left the one survey per day out because it was annoying (I doubt this will be a problem for some time - in the event it went live, I'd say Doctor's filling in surveys too frequently will be a nice problem to have and I can easily add it back).

Aborting, picking up old surveys and the business model selection is all in. There is a placeholder for the company logo so we can add that as soon as you have a few logos.

You can now review and edit the surveys much more easily (e.g.copy and paste new files into https://github.com/SHeffernan/RxLoop/tree/master/survey_populate/populations/demo/questionnaire_source basing them on e.g. https://github.com/SHeffernan/RxLoop/blob/master/survey_populate/populations/demo/questionnaire_source/default.yaml). yaml is an old serialization format which is well understood by lots of 'text' editors if you want nice colours and layout help. As you can see, github understands it too, which is nice. The questions etc. have a number that defines their order but is also used in the prerequisite identification which reads question_order/answer_order.

Notice that I changed 3/2 to use prerequisites because the question felt silly if I'd answered 'Begin Use' in 3/1.

If you add a column to the drug list with the name of the desired survey, I can very easily make the necessary associations - the drug list is built into the population source - https://github.com/SHeffernan/RxLoop/blob/master/survey_populate/populations/demo/drug_list.csv.

I also twinked admin a bit so it is easier to use.
